Add deny policies

aanm opened the pull request:
Deny rules will be configured with `ingressDeny` and `egressDeny`.

Policy example:
```yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "guestbook-web"
spec:
  endpointSelector:
    matchLabels:
      k8s-app.guestbook: web
  ingress:
  - fromEndpoints:
    - matchLabels:
        k8s-app.guestbook: redis
  ingressDeny:
  - fromEndpoints:
    - matchLabels:
        k8s-app.guestbook: redis
```

Since in the datapath the _denies_ take priority over the _allow_ rules, this rule will drop all packets from any pod running with the label `k8s-app.guestbook=redis` in the default namespace that tries to reach any pod running with the label `k8s-app.guestbook=web`.

The `ingressDeny` differs from the `ingress` as the L7 rules are not available in the L4 section. For example:
`ingress`:
```yaml
...
  ingress:
  - fromEndpoints:
    - matchLabels:
        env: prod
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/public"
...
```
vs
`ingressDeny`:
```yaml
...
  ingressDeny:
  - fromEndpoints:
    - matchLabels:
        env: prod
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
...
```
And in `egressDeny` vs `egress` the `toFQDNs` are not available as well as, similar to the `ingressDeny`, the L7 Rules.

`egress`:
```yaml
...
  egress:
  - toEndpoints:
    - matchLabels:
       "k8s:io.kubernetes.pod.namespace": kube-system
       "k8s:k8s-app": kube-dns
    toPorts:
      - ports:
         - port: "53"
           protocol: ANY
        rules:
          dns:
            - matchPattern: "*"
  - toFQDNs:
      - matchName: "cilium.io"
      - matchName: "sub.cilium.io"
      - matchPattern: "*.sub.cilium.io"
...
```
`egressDeny`:
```yaml
...
  egressDeny:
  - toEndpoints:
    - matchLabels:
       "k8s:io.kubernetes.pod.namespace": kube-system
       "k8s:k8s-app": kube-dns
    toPorts:
      - ports:
         - port: "53"
           protocol: ANY
...
```


TODO:
- [x] Fix tests
- [x] Add tests
- [x] Documentation
- [x] Change the slightly the exposed API to the users since deny policies won't have L7 policies initially.
     This was done in the last commit if we don't like it we can always revert it.

Fixes https://github.com/cilium/cilium/issues/7111

aanm committed:
k8s/apis: refactor parseToCiliumRule functions

This will allow to reuse functions.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/policy: add function to retrieve identities of policyMapState

Create function to de-duplicate code and have the ability to re-use
code.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint: remove unused functions

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v1: add policy deny specific fields

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
cilium/cmd: dump policy deny maps

With these changes users will have the ability to dump deny
policy maps using 'cilium bpf policy map'.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
bpf: add deny policy enforcement

This change adds the deny policy enforcement in the datapath. When
checking the entry of the policy map, we will check the dedicated bit
for policy denies. If that bit is set, which we will consider unlikely
to avoid performance penalties for the allow case, the traffic will be
dropped.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
policy/api: add IngressDeny and EgressDeny fields to API

These changes only add the fields into the API for both Cilium and
Kubernetes interactions.

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme commented on bpf/lib/policy.h:
```diff
@@ -190,6 +190,14 @@ policy_can_access_ingress(struct __ctx_buff *ctx, __u32 srcID, __u32 dstID,
+				      proto, CT_INGRESS, is_untracked_fragment,
+                                      match_type);
+        if (ret >= 0) {
+	        cilium_dbg(ctx, DBG_POLICY_DENIED, srcID, dstID);
```
These lines have some indentation issues (spaces instead of tabs?).

The test on the return value deserves a comment. On the allow map below the test is `(ret >= CTX_ACT_OK)` for the allow case (ret can be a proxy port or 0). Here the same test is used for the deny case, but it is not readily obvious since the difference between using `0` here and `CTX_ACT_OK` below.

jrajahalme commented on bpf/lib/policy.h:
```diff
@@ -229,6 +237,13 @@ policy_can_egress(struct __ctx_buff *ctx, __u32 srcID, __u32 dstID,
+	ret = __policy_can_access(&POLICY_DENY_MAP, ctx, srcID, dstID, dport, proto,
+				  CT_EGRESS, false, match_type);
+        if (ret >= 0) {
+	        cilium_dbg(ctx, DBG_POLICY_DENIED, srcID, dstID);
```
Same (whitespace)

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -210,6 +210,7 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 	updatedDesiredMapState := make(policy.MapState)
 
 	for _, l4 := range m {
+		// Deny policies do not support redirects
```
👍

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1192,12 +1211,37 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 	if errors > 0 {
 		return proxyChanges, fmt.Errorf("updating desired PolicyMap state failed")
-	} else if len(adds)+len(deletes) > 0 {
+	} else if len(allowAdds)+len(allowDeletes) > 0 {
```
should count denies here as well

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1192,12 +1211,37 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 
+	for keyToAdd, entry := range denyAdds {
+		// Keep the existing proxy port, if any
+		if entry.IsRedirectEntry() {
```
Deny should never be a redirect entry, so maybe flag an error here if so?

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1229,6 +1273,17 @@ func (e *Endpoint) syncPolicyMap() error {
+			// If key that is in realized state is not in desired state, just remove it.
+			if _, ok := e.desiredPolicy.PolicyDenyMapState[keyToDelete]; !ok {
+				if deleted, err := deletePolicyKey(e.policyDenyMap, e.realizedPolicy.PolicyDenyMapState, e.desiredPolicy.PolicyDenyMapState, keyToDelete, false, nil); !deleted {
+					e.getLogger().WithError(err).Error("Failed to delete PolicyMap key")
```
update log message to signify deny case. Also maybe consider refactoring to remove code duplication?

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1262,6 +1317,17 @@ func (e *Endpoint) addPolicyMapDelta() error {
+			if added, err := addPolicyKey(e.policyDenyMap, e.realizedPolicy.PolicyDenyMapState, e.desiredPolicy.PolicyDenyMapState, keyToAdd, entry, false); !added {
+				e.getLogger().WithError(err).WithFields(logrus.Fields{
+					logfields.Port: entry.ProxyPort,
+				}).Error("Failed to add PolicyMap key")
```
message detail (add "deny") and possible refactor?

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1343,6 +1428,35 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+		if _, ok := e.desiredPolicy.PolicyDenyMapState[keyToDelete]; !ok {
+			e.getLogger().WithField(logfields.BPFMapKey, entry.String()).Debug("syncPolicyMapWithDump removing a bpf policy entry not in the desired state")
+			if deleted, err := deletePolicyKey(e.policyDenyMap, e.realizedPolicy.PolicyDenyMapState, e.desiredPolicy.PolicyDenyMapState, keyToDelete, false, nil); !deleted {
+				e.getLogger().WithError(err).Error("Failed to delete PolicyMap key")
```
log message + potential for more refactoring to remove code duplication?

jrajahalme reviewed (commented):
Looking good, some minor comments only so far. Overall one would hope that most of the code duplication for the deny case could be avoided by refactoring code.

aanm committed:
pkg/{endpoint,policy}: add policy deny implementation

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/policy: use PortDenyRules for deny policies

Since not all functionalities will be available in the IngressDeny and
EgressDeny fields of Rule, it was necessary to create a new structure,
IngressDeny and EgressDeny. These 2 structures share common fields with
the Ingress and Egress except the implementation of L7 Policies, which
is part of the PortRule but not of the PortDenyRules. This requirement
requires the introduction of some interfaces to avoid code duplication
in the policy calculation.

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -248,6 +248,7 @@ func (e *Endpoint) getEndpointPolicy() (policy *cilium_v2.EndpointPolicy) {
 		// If either ingress or egress policy is enabled, go through
 		// the desired policy to populate the values.
 		if !allowsAllIngress || !allowsAllEgress {
+			// TODO add for deny policies
```
Do you plan to do this for in this PR?

aanm replied:
Ah yes. I forgot about this :smile:

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -249,18 +276,33 @@ func (p *Repository) AllowsIngressRLocked(ctx *SearchContext) api.Decision {
+				}
+			}
+		} else if len(ingressPolicy) > 0 {
+			verdict = ingressPolicy.IngressCoversContext(ctx)
```
Normal policy should still be considered if verdict after checking deny policies is not `api.Denied`, so this can not be in an `else` block.

aanm replied:
nice catch!

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -283,17 +325,32 @@ func (p *Repository) AllowsEgressRLocked(ctx *SearchContext) api.Decision {
+				}
+			}
+		} else if len(egressPolicy) > 0 {
+			verdict = egressPolicy.EgressCoversContext(ctx)
```
Same here.

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -531,13 +588,14 @@ func (p *Repository) getMatchingRules(securityIdentity *identity.Identity) (ingr
 				egressMatch = true
 			}
 			matchingRules = append(matchingRules, r)
+			// matchingDenyRules = append(matchingDenyRules, r)
```
Why this in comments?

aanm replied:
leftovers

jrajahalme commented on pkg/policy/resolve.go:
```diff
@@ -114,14 +124,17 @@ func (p *selectorPolicy) Detach() {
 		calculatedPolicy.PolicyMapState.AllowAllIdentities(
 			!p.IngressPolicyEnabled, !p.EgressPolicyEnabled)
+		// Do not "allow" any identity in the PolicyMapDeny. Doing it so it will
+		// mean the traffic will be blocked
```
I take this comment explains why there is no "corresponding" call with deny map state. IMO this should be placed before the call and expressed in terms of why only "PolicyMapState" is used here.

jrajahalme commented on pkg/policy/resolve.go:
```diff
@@ -142,6 +155,8 @@ func (p *selectorPolicy) DistillPolicy(policyOwner PolicyOwner, isHost bool) *En
 	if !isHost {
 		calculatedPolicy.PolicyMapState.DetermineAllowLocalhostIngress(p.L4Policy)
+		// Do not "allow" any identity in the PolicyMapDeny. Doing it so it will
+		// mean the traffic will be blocked
```
Same here.

jrajahalme commented on pkg/policy/rule.go:
```diff
@@ -512,6 +535,23 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
+			found += cnt
+		}
+	}
+
```
I wonder if `SetDeny(true)` needs to be called before this block as well?

jrajahalme commented on pkg/policy/rule.go:
```diff
@@ -530,12 +570,26 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
+			found += cnt
+		}
+	}
+
```
I wonder if `SetDeny(true)` needs to be called before this block as well?

jrajahalme reviewed (commented):
Some questions for clarification. Also, it's a pity that the last commit makes code less readable :-(

aanm committed:
refactor desiredPolicyAllowsIdentity out of pkg/endpoint

This function can be refactored out of the pkg/endpoint. With this
change it is also possible to unit test this function.

Signed-off-by: André Martins <andre@cilium.io>

pchaigno commented on bpf/lib/maps.h:
```diff
@@ -65,6 +65,18 @@ struct bpf_elf_map __section_maps POLICY_MAP = {
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= POLICY_MAP_SIZE,
+	.flags		= CONDITIONAL_PREALLOC,
+};
```
I'm assuming this will require a note in the upgrade guide since it increases per-endpoint memory consumption by ~0.5Mo.

aanm replied:
I'm not sure TBH. Do we have a minimal memory requirement per endpoint in our documentation?

pchaigno replied:
I couldn't find one. I'm guessing users will monitor memory usage on nodes and I'm wondering if the increase will be large enough that they would take notice. If we think that's possible, we'd better have a note in the guide. I guess that mostly depends on how many containers per node we can expect in the largest clusters.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -190,6 +190,18 @@ policy_can_access_ingress(struct __ctx_buff *ctx, __u32 srcID, __u32 dstID,
+	 *  and so we should drop the packets
+	 */
+	if (ret >= CTX_ACT_OK) {
+		cilium_dbg(ctx, DBG_POLICY_DENIED, srcID, dstID);
```
Do we need to extend `match_type` and the userspace code to properly signal deny policies in `cilium monitor`'s policy verdicts? I'm afraid the current output could be confusing to users.

Extending `match_type` could simply be a matter of changing the current `enum` to flags:
```c
enum {
	POLICY_MATCH_NONE = 0,
	POLICY_MATCH_L3_ONLY = 1,
	POLICY_MATCH_L3_L4 = 2,
	POLICY_MATCH_L4_ONLY = 3,
	POLICY_MATCH_ALL = 4,
};
```
would become
```c
#define POLICY_MATCH_NONE 0
#define POLICY_MATCH_L3 1
#define POLICY_MATCH_L3_ONLY POLICY_MATCH_L3
#define POLICY_MATCH_L4 2
#define POLICY_MATCH_L4_ONLY POLICY_MATCH_L4
#define POLICY_MATCH_L3_L4 (POLICY_MATCH_L3|POLICY_MATCH_L4)
#define POLICY_MATCH_ALL 4
#define POLICY_MATCH_DENY 8
```

pchaigno commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -31,7 +31,7 @@ const (
 	// CustomResourceDefinitionSchemaVersion is semver-conformant version of CRD schema
 	// Used to determine if CRD needs to be updated in cluster
-	CustomResourceDefinitionSchemaVersion = "1.22"
+	CustomResourceDefinitionSchemaVersion = "1.22.1"
```
Why is this updated to `1.22.1` instead of `1.23`?

aanm replied:
Is still WIP but it's from https://github.com/cilium/cilium/pull/12493

pchaigno commented on pkg/policy/mapstate.go:
```diff
@@ -137,24 +137,36 @@ func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry) {
+			delete(keys, localHostKey)
+			if !option.Config.EnableRemoteNodeIdentity {
+				delete(keys, localRemoteNodeKey)
+			}
```
AFAIU, that will only remove L3-only rules for traffic from the local host. What if there is a L4 or L3+L4 deny rule?

By the way, how can someone add a L4-only rule without blocking access from the local host...?

aanm replied:
>AFAIU, that will only remove L3-only rules for traffic from the local host. What if there is a L4 or L3+L4 deny rule?

How is that different than the way we currently doing it for the non-deny policies?

pchaigno replied:
With allow policies, adding a new L3+L4 rule for traffic from the host won't overwrite the existing allow-all-from-host rule:
```
$ tail -n8 test.yaml 
  ingress:
  - fromEntities:
    - host
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
$ kubectl create -f test.yaml
$ sudo cilium bpf policy get 2880 | grep Ingress
Ingress     reserved:host                                      80/TCP       NONE         0        0         
Ingress     reserved:host                                      ANY          NONE         13206    93
```
I don't think users can break ingress from host to pod with allow policies. But it looks like they can with deny policies, as follows (not tested):
```yaml
  ingressDeny:
  - fromEntities:
    - host
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
```
As far as I understand, even the following would break our existing allow-all-from-host:
```yaml
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
```

pchaigno commented on pkg/policy/repository.go:
```diff
@@ -58,11 +58,23 @@ type PolicyContext interface {
 	GetEnvoyHTTPRules(l7Rules *api.L7Rules) (*cilium.HttpNetworkPolicyRules, bool)
+
+	// IsDeny returns true if the policy computation should be done for the
+	// policy deny case. This function return different values depending on the
```
```suggestion
	// policy deny case. This function returns different values depending on the
```

pchaigno reviewed (changes requested):
LGTM overall. My only concern is around handling of `AlwaysAllowLocalhost` when deny policy rules are loaded, in particular with L4-only deny rules (see comment below). I think all other comments can be addressed as follow ups.

Since L7 rules are not supported in deny rules, should we reject them during validation?

That last commit is really hard to review. Couldn't it be broken up in several commits?

christarazi commented on pkg/policy/resolve_test.go:
```diff
@@ -660,3 +661,88 @@ func (ds *PolicyTestSuite) TestMapStateWithIngress(c *C) {
+			},
+			wantIngress: true,
+			wantEgress:  false,
+		},
```
Are we missing a test case here for "policy enabled for egress"? We have one for "policy enabled for ingress".

christarazi commented on pkg/policy/mapstate.go:
```diff
@@ -185,6 +185,32 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
 
+// GetIdentities returns the ingress and egress identities stored in the
+// MapState.
+func (pms MapState) GetIdentities(log *logrus.Logger) (ingIdentities, egIdentities []int64) {
```
Nit:
This is using named returns, but the return statement still references the variables. We can either remove the named return or don't reference the variables when returning.

jrajahalme replied:
I don't know if we have an official style to follow here, but I find that sometimes named return variables are handy for documenting them, also avoids adding initializers. Depending on the complexity of the function, I still prefer returning the variables explicitly.

christarazi commented on cilium/cmd/bpf_policy_get.go:
```diff
@@ -83,11 +85,24 @@ func listAllMaps() {
+		fmt.Println("POLICY DENY MAP")
+		denyMapPath, err := endpointToDenyPolicyMapPath(epID)
+		if err != nil {
+			Fatalf("Failed to parse endpointID %q", epID)
```
Nit: 

```suggestion
			Fatalf("Failed to parse endpoint ID %q", epID)
```

christarazi commented on cilium/cmd/bpf_policy_get.go:
```diff
@@ -83,11 +85,24 @@ func listAllMaps() {
+	mapPath, err := endpointToPolicyMapPath(epID)
 	if err != nil {
-		Fatalf("Failed to parse endpointID %q", lbl)
+		Fatalf("Failed to parse endpointID %q", epID)
```
```suggestion
		Fatalf("Failed to parse endpoint ID %q", epID)
```

christarazi commented on pkg/policy/resolve.go:
```diff
@@ -183,18 +203,20 @@ func (p *EndpointPolicy) computeDirectionL4PolicyMapEntries(l4PolicyMap L4Policy
 // locking the selector cache to make sure concurrent identity updates
 // have completed.
-func (p *EndpointPolicy) ConsumeMapChanges() (adds, deletes MapState) {
+func (p *EndpointPolicy) ConsumeMapChanges() (allowAdds, allowDeletes, denyAdds, denyDeletes MapState) {
```
Would it make sense to convert the returns into a struct type instead?

christarazi commented on pkg/policy/rule.go:
```diff
@@ -512,6 +535,23 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
 
+	for _, ingressRule := range r.IngressDeny {
+		// TODO (ianvernon): GH-1658
+		var allCIDRs []api.CIDR
```
I think we can preallocate the minimum size of this slice here:

```suggestion
		allCIDRs := make([]api.CIDR, 0, len(ingressRule.FromCIDR))
```

christarazi commented on pkg/policy/rule.go:
```diff
@@ -530,12 +570,26 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
 	}
 
+	for _, egressRule := range r.EgressDeny {
+		var allCIDRs []api.CIDR
```
Preallocate slice here as well

christarazi commented on pkg/policy/api/l4.go:
```diff
@@ -156,6 +156,36 @@ type PortRule struct {
+type PortDenyRule struct {
+	// Ports is a list of L4 port/protocol
+	//
+	// +optional
```
I believe we've converted all of `+optional` to `+kubebuilder:validation:Optional`.

christarazi commented on pkg/policy/api/l4.go:
```diff
@@ -199,3 +229,58 @@ func (rules *L7Rules) Len() int {
 	return rules == nil || (rules.HTTP == nil && rules.Kafka == nil && rules.DNS == nil && rules.L7 == nil)
 }
+
+// PortRules is a slice of PortRule
```
```suggestion
// PortRules is a slice of PortRule.
```

These fields potentially may show in the CRD, and this description will be embedded there. We should be consistent with our formatting.

christarazi commented on pkg/policy/api/l4.go:
```diff
@@ -199,3 +229,58 @@ func (rules *L7Rules) Len() int {
+	return len(pr)
+}
+
+// PortDenyRules is a slice of PortDenyRule
```
Same here with formatting.

christarazi reviewed (changes requested):
First-pass looks good overall. Some items to address.

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -190,6 +190,18 @@ policy_can_access_ingress(struct __ctx_buff *ctx, __u32 srcID, __u32 dstID,
 
+	ret = __policy_can_access(&POLICY_DENY_MAP, ctx, dstID, srcID, dport,
+					proto, CT_INGRESS, is_untracked_fragment,
+					match_type);
```
Checkpatch reported an issue here:
```
 CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
#73: FILE: bpf/lib/policy.h:194:
+	ret = __policy_can_access(&POLICY_DENY_MAP, ctx, dstID, srcID, dport,
+					proto, CT_INGRESS, is_untracked_fragment,
```

aanm committed:
pkg/policy: add unit tests for deny policies

Signed-off-by: André Martins <andre@cilium.io>

christarazi commented on pkg/policy/resolve_test.go:
```diff
@@ -660,3 +661,88 @@ func (ds *PolicyTestSuite) TestMapStateWithIngress(c *C) {
+			},
+			wantIngress: true,
+			wantEgress:  false,
+		},
```
Missing `policy enabled for egress` case?

christarazi commented on api/v1/models/endpoint_policy.go:
```diff
@@ -35,6 +35,14 @@ type EndpointPolicy struct {
 	// cidr policy
 	CidrPolicy *CIDRPolicy `json:"cidr-policy,omitempty"`
 
+	// List of identities to which this endpoint is not allowed to communicate
```
```suggestion
	// List of identities to which this endpoint is not allowed to communicate to
```

joestringer replied:
The `to which` already has the `to`, but it's it /is/ easy to miss given that this is a slightly unusual sentence structure.

christarazi commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -31,7 +31,7 @@ const (
 	// CustomResourceDefinitionSchemaVersion is semver-conformant version of CRD schema
 	// Used to determine if CRD needs to be updated in cluster
-	CustomResourceDefinitionSchemaVersion = "1.22.1"
+	CustomResourceDefinitionSchemaVersion = "1.22.2"
```
Do we need to bump the schema again during 1.9 dev cycle?

aanm replied:
Yes because we have made a `1.9.0-rc.0` which contains `1.22.1`

christarazi commented on pkg/policy/api/rule.go:
```diff
@@ -57,12 +57,28 @@ type Rule struct {
+	// rules in the 'ingress' field.
+	// If omitted or empty, this rule does not apply at ingress.
+	//
+	// +optional
```
Nit: 
```suggestion
	// +kubebuilder:validation:Optional
```

For the sake of consistency.

christarazi commented on examples/crds/ciliumendpoints.yaml:
```diff
@@ -242,7 +242,7 @@ spec:
                     properties:
                       adding:
-                        description: AllowedIdentityList is a list of AllowedIdentityTuple.
+                        description: Deprecated
```
It would have been nice if K8s allowed you to mark a field as deprecated. I think newer versions of K8s-provided types can give deprecation warnings, but that doesn't seem to be exposed to CRDs.

christarazi commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+
+	"github.com/cilium/cilium/pkg/checker"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/api"
```
The ordering of these should be swapped.

christarazi commented on pkg/policy/resolve_deny_test.go:
```diff
@@ -0,0 +1,513 @@
+		}
+		rule.Sanitize()
+		rules = append(rules, &rule)
+	}
```
```suggestion
	rules := make([]api.Rules, 0, numRules)
	for i := 1; i <= numRules; i++ {
		rule := api.Rule{
			EndpointSelector: fooSelector,
			IngressDeny:      []api.IngressDenyRule{ingDenyRule},
		}
		rule.Sanitize()
		rules = append(rules, &rule)
	}
```

christarazi replied:
I see that this loop is repeated, so might be worth extracting out.

aanm replied:
One is for ingress the other for egress.

christarazi reviewed (changes requested):
First pass through. Ignoring some of the commented out code and TODOs.

jrajahalme commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -109,108 +109,111 @@ func getEndpointSelector(namespace string, labelSelector *slim_metav1.LabelSelec
 
-func parseToCiliumIngressRule(namespace string, inRule, retRule *api.Rule) {
-	matchesInit := matchesPodInit(retRule.EndpointSelector)
+func parseToCiliumIngressRule(namespace string, es api.EndpointSelector, inRules []api.IngressRule) []api.IngressRule {
```
nit: should rename to `parseToCiliumIngressRules()` or `copyIngressRules()`? (I don't see much parsing taking place, also the return type is the same as input type)

aanm replied:
That's a good point. I'll leave it as is since the caller of this function is also named `ParseToCiliumRule`.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -1053,7 +1053,16 @@ func (e *Endpoint) SkipStateClean() {
 // The bool pointed by hadProxy, if not nil, will be set to 'true' if
 // the deleted entry had a proxy port assigned to it.  *hadProxy is
 // not otherwise changed (e.g., it is never set to 'false').
-func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, hadProxy *bool) bool {
```
nit: Call sites would keep simpler if in addition to `deletePolicyKey()` you'd keep `Endpoint.deletePolicyKey()` as the latter could derive parameters to the former from the Endpoint?

aanm replied:
I've dropped this commit.

jrajahalme commented on Documentation/cmdref/cilium_bpf_policy_add.md:
````diff
@@ -15,6 +15,7 @@ cilium bpf policy add <endpoint id> <traffic-direction> <identity> [port/proto]
 ### Options
 
 ```
+      --deny   Sets deny mode
````
Is this still needed?

jrajahalme commented on Documentation/cmdref/cilium_bpf_policy_delete.md:
````diff
@@ -15,6 +15,7 @@ cilium bpf policy delete <endpoint id> <identity> [port/proto] [flags]
 ### Options
 
 ```
+      --deny   Sets deny mode
````
Is this still needed?

jrajahalme commented on cilium/cmd/bpf_policy_add.go:
```diff
@@ -20,17 +20,20 @@ import (
 }
 
 func init() {
+	bpfPolicyAddCmd.Flags().BoolVar(&isDeny, "deny", false, "Sets deny mode")
```
Is this still needed?

aanm replied:
Yes, otherwise there won't be a way to set the deny bit for the entry that we are manually adding with the CLI

jrajahalme commented on cilium/cmd/bpf_policy_delete.go:
```diff
@@ -27,10 +27,11 @@ var bpfPolicyDeleteCmd = &cobra.Command{
 }
 
 func init() {
+	bpfPolicyDeleteCmd.Flags().BoolVar(&isDeny, "deny", false, "Sets deny mode")
```
Is this still needed?

aanm replied:
Yes, otherwise there won't be a way to set the deny bit for the entry that we are manually adding with the CLI

joestringer replied:
This one is a delete command, is it checking for this bit and only deleting the entry if the bit is set? Or does it work another way?

aanm replied:
>This one is a delete command, is it checking for this bit and only deleting the entry if the bit is set? Or does it work another way?

yes, it's exactly like that, it will check if the deny bit is set.

joestringer replied:
OK, then `Sets deny mode` doesn't seem like a very helpful description of the flag. Maybe
```suggestion
	bpfPolicyDeleteCmd.Flags().BoolVar(&isDeny, "deny", false, "Only delete deny policy entries")
```

jrajahalme commented on pkg/maps/policymap/policymap.go:
```diff
@@ -202,24 +267,32 @@ func newKey(id uint32, dport uint16, proto u8proto.U8proto, trafficDirection tra
 // `trafficDirection` for identity `id` with destination port `dport` over
 // protocol `proto`. It is assumed that `dport` and `proxyPort` are in host byte-order.
-func (pm *PolicyMap) Allow(id uint32, dport uint16, proto u8proto.U8proto, trafficDirection trafficdirection.TrafficDirection, proxyPort uint16) error {
+func (pm *PolicyMap) Allow(id uint32, dport uint16, proto u8proto.U8proto, trafficDirection trafficdirection.TrafficDirection, proxyPort uint16, isDeny bool) error {
```
Maybe define another function (`Deny()`) instead of changing this?

jrajahalme commented on pkg/maps/policymap/policymap.go:
```diff
@@ -202,24 +267,32 @@ func newKey(id uint32, dport uint16, proto u8proto.U8proto, trafficDirection tra
 // Returns an error if the update of the PolicyMap fails.
-func (pm *PolicyMap) AllowKey(k PolicyKey, proxyPort uint16) error {
-	return pm.Allow(k.Identity, k.DestPort, u8proto.U8proto(k.Nexthdr), trafficdirection.TrafficDirection(k.TrafficDirection), proxyPort)
+func (pm *PolicyMap) AllowKey(k PolicyKey, proxyPort uint16, isDeny bool) error {
```
Maybe define another function (DenyKey()) instead of changing this?

jrajahalme commented on bpf/lib/common.h:
```diff
@@ -379,6 +379,7 @@ enum {
 #define DROP_PROXY_LOOKUP_FAILED	-178
 #define DROP_PROXY_SET_FAILED	-179
 #define DROP_PROXY_UNKNOWN_PROTO	-180
+#define DROP_POLICY_DENY	-181
```
The commit message still refers to "deny map", needs to be reworded. (not related to this line, but can't add review comments to commit messages :-)

jrajahalme commented on examples/crds/ciliumendpoints.yaml:
```diff
@@ -277,10 +277,28 @@ spec:
+                      denied:
+                        description: AllowedIdentityList is a list of AllowedIdentityTuple.
+                        items:
+                          description: AllowedIdentityTuple specifies an allowed peer by identity, destination port and protocol.
```
It's confusing to see "Allowed" in the descriptions of `denied:`

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -333,11 +334,11 @@ func (e *Endpoint) addVisibilityRedirects(ingress bool, desiredRedirects map[str
 		visPolicy = e.visibilityPolicy.Ingress
 		direction = trafficdirection.Ingress
-		policyEnabled = e.desiredPolicy.IngressPolicyEnabled
+		policyEnabled = e.desiredPolicy.IngressPolicyEnabled || e.desiredPolicy.IngressDenyPolicyEnabled
```
Why we need to have both `IngressPolicyEnabled` and `IngressDenyPolicyEnabled`? Could we just have `IngressPolicyEnabled` covering both?

aanm replied:
@jrajahalme because of [[0]](https://github.com/cilium/cilium/pull/12716/commits/0c679462cc925ac0fe9d76744a55285421c42a46#r492670095)

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -333,11 +334,11 @@ func (e *Endpoint) addVisibilityRedirects(ingress bool, desiredRedirects map[str
 		visPolicy = e.visibilityPolicy.Egress
 		direction = trafficdirection.Egress
-		policyEnabled = e.desiredPolicy.EgressPolicyEnabled
+		policyEnabled = e.desiredPolicy.EgressPolicyEnabled || e.desiredPolicy.EgressDenyPolicyEnabled
```
same for egress?

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -102,12 +104,14 @@ func NewMapStateEntry(derivedFrom labels.LabelArrayList, redirect bool) MapState
+// Note to reviewers: Should I rename this function? Name proposals?
 func (e *MapStateEntry) IsRedirectEntry() bool {
-	return e.ProxyPort != 0
+	return !e.IsDeny && e.ProxyPort != 0
```
nit: This is unnecessary change, deny entries will never have `ProxyPort` set so it suffices to test the `ProxyPort` to figure out if the entry is a redirect entry.

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -116,7 +120,51 @@ func (e *MapStateEntry) Equal(o *MapStateEntry) bool {
+			for k, v := range keys {
+				if k.TrafficDirection == key.TrafficDirection &&
+					k.Identity == key.Identity {
+					if !v.IsDeny {
```
Could also delete all prior deny keys if the new key is L3-only, as in that case the L3/L4 keys for the same identity and direction are not needed.

aanm replied:
Good point!

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -620,15 +668,20 @@ func (p *Repository) resolvePolicyLocked(securityIdentity *identity.Identity) (*
+		IngressPolicyEnabled:     ingressEnabled,
+		EgressPolicyEnabled:      egressEnabled,
+		IngressDenyPolicyEnabled: ingressDenyEnabled,
+		EgressDenyPolicyEnabled:  egressDenyEnabled,
```
Do we still need the extra fields for deny as there are no separate maps any more?

aanm replied:
I'll wait for Joe's input.

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -652,7 +705,7 @@ func (p *Repository) resolvePolicyLocked(securityIdentity *identity.Identity) (*
 	}
 
-	if ingressEnabled {
+	if ingressEnabled || ingressDenyEnabled {
```
Feels like we do not need a separate `ingressDenyEnabled` or `egressDenyEnabled` any more?

aanm replied:
I'll wait for Joe's input.

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -680,6 +733,7 @@ func (p *Repository) resolvePolicyLocked(securityIdentity *identity.Identity) (*
 
 		calculatedPolicy.CIDRPolicy.Egress = newCIDREgressPolicy.Egress
 		calculatedPolicy.L4Policy.Egress = newL4EgressPolicy
+
```
nit: extra line added here?

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -693,34 +747,38 @@ func (p *Repository) resolvePolicyLocked(securityIdentity *identity.Identity) (*
+	ingress, egress bool,
+	ingressDeny, egressDeny bool,
+	matchingRules ruleSlice,
+) {
```
Do we still need this change?

aanm replied:
I'll wait for Joe's input.

jrajahalme commented on pkg/policy/resolve.go:
```diff
@@ -121,8 +129,8 @@ func (p *selectorPolicy) DistillPolicy(policyOwner PolicyOwner, isHost bool) *En
-		calculatedPolicy.PolicyMapState.AllowAllIdentities(
-			!p.IngressPolicyEnabled, !p.EgressPolicyEnabled)
+		// TODO(aanm) Note to reviewers, what about the deny policies?
+		calculatedPolicy.PolicyMapState.AllowAllIdentities(!p.IngressPolicyEnabled, !p.EgressPolicyEnabled)
```
This TODO would be resolved if we remove support for separately enabling Deny policies?

aanm replied:
I believe so! Let's wait for Joe's input.

jrajahalme commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -109,107 +109,156 @@ func getEndpointSelector(namespace string, labelSelector *slim_metav1.LabelSelec
 			}
-
+			retRules[i].IngressCommonRule = parseToCiliumIngressCommonRule(namespace, es, ing.IngressCommonRule)
 			retRules[i].SetAggregatedSelectors()
```
Should aggregation be done on the Common function instead of all the callers?

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+}
+
+// Case 2: deny all at L3/L7 in one rule, and select an endpoint and deny all on L7
+// in another rule. Should resolve to just allowing all on L3/L7 (first rule
```
`L3/L7` here should be `L3/L4` instead (no L7 in deny rules).

aanm replied:
Yes, I need to take a 2nd look at all test case descriptions.

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+	c.Assert(state.matchedDenyRules, Equals, 0)
+}
+
+// Case 3: deny all on L7 in both rules, but select different endpoints in each rule.
```
`L4`, not `L7`

aanm replied:
Yes, I need to take a 2nd look at all test case descriptions.

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+}
+
+// Case 4: allow all at L3/L7 in one rule, and deny a selected an endpoint in
+// another rule. Should resolve to just allowing all on L3/L7 (first rule
```
`L3/L4`, not `L3/L7`

jrajahalme replied:
-> `deny a selected endpoint`

aanm replied:
Yes, I need to take a 2nd look at all test case descriptions.

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+
+// Case 4: allow all at L3/L7 in one rule, and deny a selected an endpoint in
+// another rule. Should resolve to just allowing all on L3/L7 (first rule
+// shadows the second) and denying that particular endpoint..
```
I don't see any "shadowing" in here, only straightforward translation?

aanm replied:
Yes, I need to take a 2nd look at all test case descriptions.

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+	c.Assert(state.matchedDenyRules, Equals, 0)
+}
+
+// Case 5: allow L7 in all endpoints in one rule, and deny a selected an
```
-> `deny a selected`

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+}
+
+// Case 5: allow L7 in all endpoints in one rule, and deny a selected an
+// endpoint in another rule. Should resolve to just allowing all on L3/L7 (first
```
`L4/L7`, not `L3/L7`

jrajahalme commented on pkg/policy/l4_filter_deny_test.go:
```diff
@@ -0,0 +1,676 @@
+// endpoint in another rule. Should resolve to just allowing all on L3/L7 (first
+// rule shadows the second) and denying that particular endpoint..
+func (ds *PolicyTestSuite) TestL7AllowRuleShadowedByL3DenyAll(c *C) {
+	// Case 4A: Specify WildcardEndpointSelector explicitly.
```
I don't see any shadowing here?

jrajahalme reviewed (approved):
Small nits only. However, it seems to me that this would be simplified a lot if we just added a `verdict=(allow|deny) to the existing ingress and egress policies (defaulting to allow). Most of the code changes would go away, but maybe we would loose some policy validation on k8s side as users could apply policies with both `verdict=deny` and L7 rules? Or can a CRD field validation be conditional on the value of another field?

tklauser commented on bpf/lib/common.h:
```diff
@@ -235,7 +235,9 @@ struct policy_key {
 struct policy_entry {
 	__be16		proxy_port;
-	__u16		pad0;
+	__u8		deny:1,
```
Looks like this change ended up in the wrong commit (`cilium/cmd: dump policy deny maps`). It should probably be part of `bpf: add deny policy enforcement`?

aanm replied:
Not quite, this change needs to be part of this commit or otherwise the CLI command to add and remove bpf entries wouldn't have any effect as well as cilium wouldn't run due mismatch of the bpf field "pad0" and the "deny" field in the controlplane.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1426,6 +1428,13 @@ func (e *Endpoint) GetPolicyVerdictLogFilter() uint32 {
+	}
+	if e.desiredPolicy.EgressDenyPolicyEnabled {
+		filter = (filter | 0x8)
+	}
```
[0] this @jrajahalme .

If it does not matter I'm happy to remove those fields. cc @joestringer, and @pchaigno maybe?

joestringer replied:
AFAICT this is trying to control whether we emit policy log events, based upon whether policy is applied or not.

If these deny policies put the endpoint into default deny then I see no reason to separate them out. Git blame says @lzang would be the one to ask about implications.

pchaigno replied:
If we're not enforcing any deny policies for an endpoint, then we're never going to return `DROP_POLICY_DENY` and thus we don't need any kind of filtering logic for verdicts.

tklauser commented on pkg/policy/api/egress.go:
```diff
@@ -255,3 +345,28 @@ func (e *EgressRule) CreateDerivative(ctx context.Context) (*EgressRule, error)
+	if !e.RequiresDerivative() {
+		return newRule, nil
+	}
+	newRule.ToCIDRSet = CIDRRuleSlice{}
```
nit: preallocate the slice

```suggestion
	newRule.ToCIDRSet = make(CIDRRuleSlice, 0, len(e.ToGroups)
```

tklauser commented on pkg/maps/policymap/policymap.go:
```diff
@@ -47,6 +48,40 @@ const (
+	}
+
+	return strings.Join(str, ", ")
+}
```
I think this can be simplified to:
```suggestion
func (pef PolicyEntryFlags) String() string {
	if pef.IsDeny() {
		return "Deny"
	}
	return "Allow"
}
```

tklauser commented on pkg/maps/policymap/policymap.go:
```diff
@@ -47,6 +48,40 @@ const (
+
+type PolicyEntryFlags uint8
+
+// UInt8 returns the UInt8 representation of the ServiceFlags.
```
Nit:

```suggestion
// UInt8 returns the UInt8 representation of the PolicyEntryFlags.
```

tklauser reviewed (commented):
Just small nits from a quick first pass.

tklauser commented on pkg/maps/policymap/policymap.go:
```diff
@@ -47,6 +47,35 @@ const (
+	return pef.is(policyFlagDeny)
+}
+
+// String returns the string implementation of ServiceFlags.
```
```suggestion
// String returns the string implementation of PolicyEntryFlags.
```

tklauser reviewed (approved):
Another small godoc nit, otherwise LGTM.

christarazi commented on pkg/maps/policymap/policymap.go:
```diff
@@ -47,6 +47,35 @@ const (
+type policyFlag uint8
+
+const (
+	policyFlagDeny = 1 << 0
```
nit: `iota`? Might as well, if we intend for further policy flags

christarazi commented on pkg/maps/policymap/policymap.go:
```diff
@@ -47,6 +47,35 @@ const (
+	policyFlagDeny = 1 << 0
+)
+
+type PolicyEntryFlags uint8
```
Nit: Godoc

christarazi commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -172,88 +172,123 @@ func updateLabels(allocator cache.IdentityAllocator, allowedIdentityTuple *ciliu
+				policy.Ingress.Allowed = cilium_v2.AllowedIdentityList{identityTuple}
+			} else {
+				policy.Ingress.Allowed = append(policy.Ingress.Allowed, identityTuple)
+			}
```
Nit: we don't need to negate `isDeny`. We can just invert these cases. Makes it easier to read first time without needing to do the mental computation :)

christarazi commented on pkg/policy/l4.go:
```diff
@@ -725,8 +775,14 @@ func (l4 L4PolicyMap) containsAllL3L4(labels labels.LabelArray, ports []*models.
+				return api.Denied
+			}
+			return api.Allowed
+		}
```
```suggestion
		switch {
		case matches && isDeny:
			return api.Denied
		case matches:
		        return api.Allowed
		}
```

joestringer replied:
interesting, @christarazi does Go define ordering between the overlapping switch cases here? ie if `matches` is true, then logically the second case is always true, regardless of whether the first case is true. If Go evaluates these in order then we're fine. If not, then the second case probably needs to more explicitly check `!isDeny` to be the inverse condition.

christarazi replied:
@joestringer It does :smile:! https://tour.golang.org/flowcontrol/10

christarazi replied:
You may have been thinking about the `select` block, which randomizes the case selection if all the cases are "ready" in the case of channels. I sometimes confuse them too, they look similar and have the same length :sweat_smile:

joestringer replied:
They're functionally so similar that I usually assume they're the same thing. Thanks for sharing the link, this helps a lot :)

christarazi commented on pkg/policy/mapstate.go:
```diff
@@ -85,12 +85,14 @@ type MapStateEntry struct {
 	// DerivedFromRules tracks the policy rules this entry derives from
 	DerivedFromRules labels.LabelArrayList
+
+	IsDeny bool
```
Nit: Godoc

christarazi commented on pkg/policy/mapstate.go:
```diff
@@ -181,14 +226,24 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
+		keys[keyToAdd] = NewMapStateEntry(derivedFrom, false, false)
 	}
 }
+func (pms MapState) GetIdentities(log *logrus.Logger) (ingIdentities, egIdentities []int64) {
+	return pms.getIdentities(log, false)
+}
+
+func (pms MapState) GetDenyIdentities(log *logrus.Logger) (ingIdentities, egIdentities []int64) {
+	return pms.getIdentities(log, true)
+}
```
You could simply keep `GetIdentities` and add a boolean parameter as `isDeny`, which removes the need for an extra function `GetDenyIdentities`. If there are many calls to `GetIdentities`, then this might not be worth it.

aanm replied:
I think it looks explicit the current way

christarazi commented on pkg/policy/repository.go:
```diff
@@ -532,11 +570,21 @@ func (p *Repository) getMatchingRules(securityIdentity *identity.Identity) (ingr
+				}
+				if len(r.IngressDeny) > 0 {
+					ingressDenyMatch = true
+				}
```
```suggestion
			if !ingressMatch {
				ingressMatch = len(r.Ingress) > 0
				ingressDenyMatch = len(r.IngressDeny) > 0
			}
```

christarazi commented on pkg/policy/repository.go:
```diff
@@ -532,11 +570,21 @@ func (p *Repository) getMatchingRules(securityIdentity *identity.Identity) (ingr
+				}
+				if len(r.EgressDeny) > 0 {
+					egressDenyMatch = true
+				}
```
Same as above

christarazi commented on pkg/policy/rule.go:
```diff
@@ -144,16 +144,26 @@ func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter,
 			// Merge two non-identical sets of non-nil rules
+			if l7Rules.IsDeny {
+				// if existing rule is deny then its a no-op
+				// denies takes priority over any rule.
```
Nit:
```suggestion
				// If existing rule is deny then it's a no-op.
				// Denies takes priority over any rule.
```

christarazi commented on pkg/policy/rule.go:
```diff
@@ -144,16 +144,26 @@ func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter,
+				continue
+			} else if newL7Rules.IsDeny {
+				// overwrite existing filter if the new rule is a deny case
+				// denies takes priority over any rule.
```
Nit:
```suggestion
				// Overwrite existing filter if the new rule is a deny case.
				// Denies takes priority over any rule.
```

christarazi reviewed (changes requested):
Did not review the long test code. Few more nits and small comments.

joestringer commented on bpf/lib/policy.h:
```diff
@@ -154,6 +168,8 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
 		account(ctx, policy);
+		if (unlikely(policy->deny))
+			return DROP_POLICY_DENY;
 		*match_type = POLICY_MATCH_ALL;
```
The other cases update `*match_type` before returning. I think that's likely what we want: policy verdict messages could still tell us not only that the traffic is being denied but that it's denied by L3/L4/L3+L4 entry.

joestringer commented on pkg/policy/distillery_test.go:
```diff
@@ -184,34 +184,44 @@ var (
 			WithIngressRules([]api.IngressRule{{
-			FromEndpoints: []api.EndpointSelector{allowFooL3_},
-			ToPorts:       allowAllL4_,
+			IngressCommonRule: api.IngressCommonRule{
```
Too bad you had to do this. I guess we missed `api.NewIngressRule().WithSelector(...)` type helpers that would otherwise have avoided forcing you to modify the structure in all of the tests :/

joestringer commented on cilium/cmd/helpers.go:
```diff
@@ -332,8 +335,16 @@ func updatePolicyKey(pa *PolicyUpdateArgs, add bool) {
+				err = policyMap.Allow(pa.label, pa.port, u8p, pa.trafficDirection, proxyPort)
+			} else {
+				err = policyMap.Deny(pa.label, pa.port, u8p, pa.trafficDirection)
+			}
```
The !deny check here adds an extra mental flip required for readers when this code could otherwise just say `if isDeny { policyMap.Deny(...) } else { policyMap.Allow(...) }`.

joestringer commented on pkg/envoy/server.go:
```diff
@@ -1143,7 +1143,10 @@ func getWildcardNetworkPolicyRule(selectors policy.L7DataMap) *cilium.PortNetwor
 		}
 
-		if l7 != nil {
+		if !l7.IsEmpty() {
```
Why is this change meaningful? Surely deny rules shouldn't be creating empty L7 (`PerSelectorPolicy`) objects in the map?

joestringer commented on pkg/maps/policymap/policymap.go:
```diff
@@ -202,9 +261,10 @@ func newKey(id uint32, dport uint16, proto u8proto.U8proto, trafficDirection tra
+func newEntry(proxyPort uint16, policyEntry PolicyEntryFlags) PolicyEntry {
 	return PolicyEntry{
 		ProxyPort: byteorder.HostToNetwork(proxyPort).(uint16),
+		Flags:     policyEntry.UInt8(),
```
```suggestion
func newEntry(proxyPort uint16, flags PolicyEntryFlags) PolicyEntry {
	return PolicyEntry{
		ProxyPort: byteorder.HostToNetwork(proxyPort).(uint16),
		Flags:     flags.UInt8(),
```

joestringer commented on pkg/maps/policymap/policymap.go:
```diff
@@ -219,7 +279,24 @@ func (pm *PolicyMap) AllowKey(k PolicyKey, proxyPort uint16) error {
+	return pm.Update(&key, &entry)
+}
+
+// DenyKey pushes an entry into the PolicyMap for the given PolicyKey k.
+// Returns an error if the update of the PolicyMap fails.
+func (pm *PolicyMap) DenyKey(k PolicyKey) error {
```
I'm not sure if it's a holdover from older code, but it's confusing to see mentions in packages like this of `DenyKey` when the key has no allow/deny portion to it. We should clean this up so that the usage is clear in the usage of the *entry* to hold the deny flag.

aanm replied:
>I'm not sure if it's a holdover from older code, but it's confusing to see mentions in packages like this of DenyKey when the key has no allow/deny portion to it. We should clean this up so that the usage is clear in the usage of the entry to hold the deny flag.

Can you clarify this a bit? This `DenyKey` is a function wrapper that invokes the entry with the deny set.

joestringer replied:
I assumed `DenyKey` means that the key itself has some deny property. But I think what we're trying to describe here as really `Upsert(k, deny)`. Anyway I think in context this is probably not so ambiguous, so I'll drop it.

joestringer commented on pkg/maps/policymap/policymap.go:
```diff
@@ -84,13 +113,37 @@ const SizeofPolicyKey = int(unsafe.Sizeof(PolicyKey{}))
 type PolicyEntry struct {
 	ProxyPort uint16 `align:"proxy_port"` // In network byte-order
-	Pad0      uint16 `align:"pad0"`
+	Flags     uint8  `align:"deny"`
+	Pad0      uint8  `align:"pad0"`
```
Have we been zeroing these pad fields on Cilium <= 1.8...?

I'm not sure if we will actually remove the policymap during upgrades if the map attributes match the previous release. Could be an upgrade headache if we haven't been.

aanm replied:
>Have we been zeroing these pad fields on Cilium <= 1.8...?
>
>I'm not sure if we will actually remove the policymap during upgrades if the map attributes match the previous release. Could be an upgrade headache if we haven't been.

these fields have been "Pad0" since `1.0.4`

https://github.com/cilium/cilium/commit/fdd4bc1d76131aa270dd24c8dc9e30268f220975

Do you think it's still worth worry about this?

joestringer replied:
My comment was not about the naming, it was about the content of the values we populate. I think if we declare the `PolicyEntry` in Go, it will auto-zero the fields like `Pad0` so we should be good.

joestringer commented on pkg/policy/api/egress.go:
```diff
@@ -145,32 +177,55 @@ type EgressRule struct {
+//
+// - If multiple members of the structure are specified, then all members
+//   must match in order for the rule to take effect. The exception to this
+//   rule is the ToRequires member; the effects of any Requires field in any
+//   rule will apply to all other rules as well.
```
What's the semantics of `ToRequires` here exactly? "The peer must have these labels to be denied"?

joestringer commented on pkg/policy/api/egress.go:
```diff
@@ -145,32 +177,55 @@ type EgressRule struct {
+//   rule is the ToRequires member; the effects of any Requires field in any
+//   rule will apply to all other rules as well.
+//
+// - ToEndpoints, ToCIDR, ToCIDRSet, ToEntities, ToServices and ToGroups are
```
Were `ToServices` and `ToGroups` here thought through or are we just piggybacking them off the existing code? Will we create derived policies for deny rules correctly, eg in the AWS security group derived policy case?

joestringer commented on pkg/policy/l4.go:
```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+		//    no redirect allow     redirect    allow   yes  (must redirect if L4-only redirects)
+		//    no redirect deny      redirect    allow   no   (deny takes precedence)
+		//
+		// have wildcard?        this is a L3L4 key?
```
Is this table encoded in unit tests as well?

joestringer replied:
(New version reworks this to use an actual lookup table in the code so we know the code matches the logic table that we reason about here)

joestringer commented on pkg/policy/l4.go:
```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+		//    no redirect allow     redirect    allow   yes  (must redirect if L4-only redirects)
+		//    no redirect deny      redirect    allow   no   (deny takes precedence)
+		//
+		// have wildcard?        this is a L3L4 key?
```
This "have wildcard" line doesn't seem to match up with the code any more.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
 		//
-		// have wildcard?        this is a L3L4 key?  not the "no" case?
-		if l4.wildcard != nil && cs != l4.wildcard && !(l7 != nil && wildcardL7Policy == nil) {
+		//    Current L3/L4         L4-only (if any)    Skip generating L3/L4 key
```
What do you mean "Current"?

jrajahalme replied:
The selector and rule (`cs`, `rule`) currently being iterated over in the `for` loop can only be potentially skipped if it is a L3/L4 rule, so `Current L3/L4`. `L4-only` if any, is another rule also in the filter, which can be the basis for skipping the `Current L3/L4` rule, i.e, <`cs`, `rule`>

joestringer commented on pkg/policy/l4.go:
```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+		wildcardL7PolicyIsDeny := wildcardL7Policy != nil && wildcardL7Policy.IsDeny
+
+		if l4.wildcard != nil && cs != l4.wildcard &&
+			// not the "no" case marked with *)?
```
When this comment is removed from the question it is answering, `not the "no" case` becomes unintelligible.

Surely we can just...?
```suggestion
			// Case marked with *)?
```

joestringer replied:
I realized afterwards that maybe "not the no case" is referring to the right hand side column in the table above, so maybe it's trying to capture the complexity of all of those conditions? Still not exactly sure though.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+			// not the "no" case marked with *)?
+			!(l7.IsRedirect() && !wildcardL7Policy.IsRedirect() && !wildcardL7PolicyIsDeny &&
+				// l3/l4 allow OR l4-only deny
+				(!isDeny || wildcardL7PolicyIsDeny)) {
```
Has someone reviewed this logic? I stared at the table above long enough to convince myself it's right, but I haven't stared at this code long enough to have the same effect yet.

joestringer replied:
(/cc @jrajahalme )

joestringer commented on pkg/policy/l4.go:
```diff
@@ -95,16 +95,31 @@ type PerSelectorPolicy struct {
 
 	api.L7Rules
+
+	// IsDeny is set if this L4Filter contains should be denied
+	IsDeny bool `json:",omitempty"`
```
Did you consider at all making this part of the key of the `L7DataMap` rather than part of the value? While this is technically called a `PerSelectorPolicy`, in practice it seems to have represented only L7 policies up until now (with L4 being represented via nil entries in a given map).

joestringer replied:
Another possibility would be to have the denies in a separate map under the L4Filter.

But I wouldn't consider addressing this for the current PR, we've already reasoned through the core logic with the current implementation. It should be possible to rework this later on without modifying the tests/behavior.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -116,7 +119,49 @@ func (e *MapStateEntry) Equal(o *MapStateEntry) bool {
+		return
+	}
+	// If there is a Deny that is an L3-only, all L3-L4 allows won't be
+	// accepted because a L3-Deny takes precedence.
```
This comment seems to assume that `entry` represents an L3+L4 entry at this point, but I don't follow how we determine this.

If it's an L4-only entry at this point, then I guess we implicitly follow the same logic and look up `{0, 0, 0}` in `keys` which I guess is never a deny, so then it just works for that case? The comments don't mention this could also occur though, seems like it should though.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -58,11 +58,23 @@ type PolicyContext interface {
 	// such rules being evaluated.
 	GetEnvoyHTTPRules(l7Rules *api.L7Rules) (*cilium.HttpNetworkPolicyRules, bool)
+
+	// IsDeny returns true if the policy computation should be done for the
+	// policy deny case. This function returns different values depending on the
+	// code path as it can be changed during the policy calculation.
+	IsDeny() bool
```
Have you tried `cilium policy trace` with deny policies? Does it report drops correctly?

joestringer reviewed (commented):
This is certainly a well-anticipated feature so it's exciting to see the PR maturing.

Given the size of the PR and the lack of broad context in the description (+ no overview in docs), I found it difficult to figure out what's worth spending time looking at.

I'm guessing this doesn't break any existing tests, so we at least have that reassurance. I see thousands of lines of new unit tests, which is encouraging to say that the code is working the way that you've intended for it to work, and it will continue to run that way. I appreciate the effort in testing, however to make progress I have ignored them for now, along with autogen code and k8s/CRD code. I believe we should actually review the tests because if the tests are wrong, the code will be wrong also. Do you have any practical suggestions to make this happen? 

I see one piece of feedback that you're looking for is around whether to conditionally enable/disable specifically the deny policy sections. I see no obvious reason to do so, seems like it's a holdover from a previous version of the PR. 

I looked at the API for a bit, particularly on egress side but some of the same feedback would apply to ingress. I'm not sure the degree to which you've thought through the inclusion of all of the fields there in deny rules. Obviously L7 you've thought about but I have some questions on some of the L3 fields.

Additional questions:
* Do deny policies put the endpoint into default-deny?
* Do we plan to mark this new feature as "beta"? 
* Have you had a dedicated review session for the interactions between the new policies and existing policies? (Just thinking that having two people look will make it somewhat less likely to end up with some unintended corner-case behaviour)

pchaigno commented on pkg/policy/mapstate.go:
```diff
@@ -137,16 +182,16 @@ func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry) {
 // the desiredPolicyKeys if the localhost is allowed to communicate with the
 // endpoint.
-func (keys MapState) DetermineAllowLocalhostIngress(l4Policy *L4Policy) {
+func (keys MapState) DetermineAllowLocalhostIngress() {
```
Unless I missed something, it will be up to the users to be careful not to block ingress from host with a deny policy. How can they use L4-only deny policies without breaking allow-all-from-host?
```yaml
  ingressDeny:
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
```

joestringer replied:
This is an interesting question and came up while I was syncing with Andre out-of-band too.

If the user explicitly writes a policy that says "deny from host", then it's a little bit hard to argue why the flag for automatic allow-all would take precedence. The default value of that flag allows from host, so it is likely not *explicitly* configured by the user. Presumably if the user configures such a policy, Cilium really should drop such traffic.

Are you concerned that users might accidentally write a "drop from host" rule?

pchaigno replied:
> Are you concerned that users might accidentally write a "drop from host" rule?

Yes, because any L4-only deny rule is a drop from host rule if I understand correctly.

I think this was addressed during the Cilium community meeting and we'll leave it to the user to be careful not to drop their traffic. Given the denies are explicit in the case of deny policies, it's probably fine.

aanm committed:
pkg/policy: refactor manual unit tests into automatically generated

These changes introduce automatic capability of performing unit tests
by adding an algorithm used to derive map states and test cases. This
commit keeps the same behavior used in the manually unit tests to
check if the new algorithm used to derive map states breaks the old
expectations. In follow up changes the manual tests will be removed.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/policy/l4.go:
```diff
@@ -313,6 +313,9 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
 		// - l3/l4 deny overrides l4-only allows (for that specific l3)
 		//
+		// This entire logic is needed for the line marked in **, apart from
+		// this line, this is pkg/policy/rule.goentirely an optimization.
```
nit: I'm guessing `pkg/policy/rule.goentirely` isn't a file so the comment has probably gotten munged here somehow.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -325,26 +328,34 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+
+		// eliminating the first 3 lines from the table above
+		if ruleHasWildcardCacheSelector && !isWildcardL7Rule &&
+			// eliminating the "Skip 'no'" case for the line marked with *)
```
Is this supposed to be 
```suggestion
			// eliminating the "Skip 'no'" case for the line marked with **
```
Or the comments above should be `*)`?

joestringer replied:
ugh, I stared at this for far longer than I'd like to admit and I didn't even realise there is now `#)`, `*)`, and `**`. Can we please just use numbers instead?

joestringer commented on pkg/policy/distillery_test.go:
```diff
@@ -586,6 +639,237 @@ func generateMapStates() []MapState {
+
+		"X	X	X	X	X	X	X	X	1	0	0	1", // 32
+		"X	X	X	X	X	X	X	X	1	0	0	1",
+		"X	X	X	X	0	1	0	0	1	0	0	1",
```
I don't think this line declares the behaviour we want, see https://docs.google.com/spreadsheets/d/1WANIoZGB48nryylQjjOw6lKjI80eVgPShrdMTMalLEw/edit?disco=AAAAHAuUUcE

joestringer commented on pkg/policy/l4.go:
```diff
@@ -267,15 +267,15 @@ func (l4 *L4Filter) GetPort() uint16 {
-func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirection.TrafficDirection) MapState {
-	port := uint16(l4.Port)
-	proto := uint8(l4.U8Proto)
+func (l4Filter *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirection.TrafficDirection) MapState {
```
nit, doesn't `L4filter` actually represent any filter including an L3-only one? We should really revisit this in a follow-up. Not sure `l4 -> l4filter` really helps all that much if the `l4` part is misleading/wrong.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -325,26 +328,34 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
-			!(l7.IsRedirect() && !wildcardL7Policy.IsRedirect() && !wildcardL7PolicyIsDeny &&
-				// l3/l4 allow OR l4-only deny
-				(!isDeny || wildcardL7PolicyIsDeny)) {
+		ruleHasWildcardCacheSelector := l4Filter.wildcard != nil
```
nit: `Cache` doesn't help readers to understand what's going on, it's just another unnecessary detail.

```suggestion
		ruleHasWildcardSelector := l4Filter.wildcard != nil
```

joestringer commented on pkg/policy/l4.go:
```diff
@@ -325,26 +328,34 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
 		// have wildcard?        this is a L3L4 key?
-		isDeny := l7 != nil && l7.IsDeny
+		isDenyRule := rule != nil && rule.IsDeny
 		wildcardL7PolicyIsDeny := wildcardL7Policy != nil && wildcardL7Policy.IsDeny
```
Except if it's a deny policy, it's not L7 right? :/

jrajahalme replied:
Right, better rename this as `wildcardRuleIsDeny`

joestringer commented on pkg/policy/l4.go:
```diff
@@ -325,26 +328,34 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+		if ruleHasWildcardCacheSelector && !isWildcardL7Rule &&
+			// eliminating the "Skip 'no'" case for the line marked with *)
+			!(rule.IsRedirect() && !wildcardL7Policy.IsRedirect() && !wildcardL7PolicyIsDeny) &&
+			// eliminating the "Skip no'" cases for the lines marked with #)
```
Thanks for the comments. I'm not convinced I understand this logic well enough to make any commentary, but I at least agree it matches the table.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -265,6 +265,41 @@ func mergeIngressPortProto(policyCtx PolicyContext, ctx *SearchContext, endpoint
 	key := p.Port + "/" + string(proto)
 	existingFilter, ok := resMap[key]
 	if !ok {
+		// Check if there is a L3-only filter for this protocol
```
I didn't review these parts.

joestringer reviewed (changes requested):
I think we need to fix the one case I highlighted in the tests. Rest of the comments I made are misc tidyup suggestions.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -313,38 +313,49 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
-			!(l7.IsRedirect() && !wildcardL7Policy.IsRedirect() && !wildcardL7PolicyIsDeny &&
-				// l3/l4 allow OR l4-only deny
-				(!isDeny || wildcardL7PolicyIsDeny)) {
+		ruleHasWildcardSelector := l4Filter.wildcard != nil
```
This would be better named `filterHasWildcardSelector`

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -313,38 +313,49 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
-				// l3/l4 allow OR l4-only deny
-				(!isDeny || wildcardL7PolicyIsDeny)) {
+		ruleHasWildcardSelector := l4Filter.wildcard != nil
+		isWildcardL7Rule := cs == l4Filter.wildcard
```
This would be better named `isWildcardRule`

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -300,11 +300,11 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
-	if l4.wildcard != nil {
-		wildcardL7Policy = l4.L7RulesPerSelector[l4.wildcard]
+	if l4Filter.wildcard != nil {
+		wildcardL7Policy = l4Filter.L7RulesPerSelector[l4Filter.wildcard]
```
This would better be named `wildcardRule`

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -128,38 +128,22 @@ func (e *MapStateEntry) DatapathEqual(o *MapStateEntry) bool {
-			for k := range keys {
+			for k, v := range keys {
 				if k.TrafficDirection == key.TrafficDirection &&
 					k.Identity == key.Identity {
```
This loop is only handling cases where the L3 is the same as in the L3-only case, i.e., the L3/L4 cases where the L3 is the same as in the L3-only deny rule (`key`). Especially, L4-only keys, where the L3 is zero (right?) are not handled, while it seems that for each L4-only rule, a L3/L4 deny rule would need to be added, as L4-only keys are looked up before L3-only keys?

jrajahalme replied:
Also, how are we now giving precedence to existing deny keys when adding a L3/L4 or L4-only allow key?

aanm commented on pkg/policy/rule.go:
```diff
@@ -265,6 +265,61 @@ func mergeIngressPortProto(policyCtx PolicyContext, ctx *SearchContext, endpoint
+				// 			}
+				// 		}
+				// 	}
+				// }
```
I'll remove these leftover comments

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -114,55 +114,91 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
+					l4OnlyAllows[k] = v
+				}
+			}
+			// Delete all L3-L4 if we are inserting a deny L3-only and
+			// there aren't allow L4-only for the existing deny L3-L4
```
Ah, interesting, the corresponding opposite to this is the case on lines 178-187 where if an L4 allow is inserted when there's already an L3 deny, then an L3+L4 deny must be injected. :+1: Not sure if we can make this relationship more clear somehow but looks right to me.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -114,55 +114,91 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
 			}
+		} else {
+			// If there is a Deny that is an L3-only, all L3-L4 denies won't be
+			// accepted because a L3-Deny takes precedence.
```
`won't be accepted` is a bit vague, and describing "L3 deny takes precedence" is a bit confusing since at the datapath level, the evaluation order is fixed to L3+L4 -> L4 -> L3 -> all, and this function is handling map state which will be directly checked by that datapath evaluation logic.

I think maybe it's sufficient to just use imperative language?

```suggestion
			// Do not insert 'newKey' if the map state already denies traffic
			// which is a superset of (or equal to) 'newKey'
```

Reading it this way, I think it's clearer to understand that you don't want to insert an entry that would allow traffic if the mapstate currently determines that that traffic must be denied.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -114,55 +114,91 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
-		// Found a L3-only Deny so we won't accept any L3-L4 policies
+		// Found a L3-only Deny so we won't accept any L3-L4 allow policies
 		return
 	}
```
What's different between these lines below and lines 162-171? Oh, I guess whether the new entry is a deny or an allow entry. Hm. Possible opportunity for code sharing?

aanm replied:
It is tricky to share the code across both branches. I'm open to suggestions!

joestringer reviewed (commented):
Minor comments on the mapstate bits. l4.go seems consistent with out-of-band review we did earlier with Jarno. :+1:  I didn't review distillery_test.go or rule.go.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -39,6 +39,70 @@ import (
+		// - l4-only deny overrides l3/l4 allows (and denies)
+		// - l3/l4 deny overrides l4-only allows (for that specific l3)
+		//
+		// This entire logic is needed for the line marked in 3), apart from
```
The reference "3)" no longer exists (should be to `(must redirect if L4-only redirects)`)

Also, the two `yes  (deny takes precedence)` cases are not optimizations, they are functionally needed. In fact the only optimizations are the `(same effect)` lines.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -299,52 +363,30 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+
+		if (l4Filter.Port != 0 || l4Filter.PortName != "") && l4Filter.wildcard != nil {
+			// Now that we have a port number and we have a wildcard therefore
+			// the filter has a L4-only rule.
```
Reword the first line of the comment as:
`// Now that we have a port number and a wildcard`

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -361,10 +403,17 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
+				logger.WithFields(logrus.Fields{
+					logfields.EndpointSelector: cs,
+					logfields.PolicyID:         identities,
+				}).Debug("ToMapState: Allowed remote IDs")
```
This is the deny case, so the comments should be swapped

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -114,55 +114,92 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
+		keys[newKey] = newEntry
+		return
+	} else if newKey.Identity == 0 {
+		// case for an existing deny L3-only and we are inserting allow L4-only
```
Here we could also be inserting a wildcard entry (allow all on all ports). Maybe should add
```
&& newKey.DestPort != 0
```

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -114,55 +114,92 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
-	key.DestPort = 0
-	key.Nexthdr = 0
-	v, ok := keys[key]
+	// branch for adding a new allow L3-L4
```
Here `DestPort` can also be 0, but seems OK.

jrajahalme reviewed (approved):
Small nits only

pchaigno commented on pkg/monitor/api/drop.go:
```diff
@@ -83,6 +83,7 @@ var errors = map[uint8]string{
 	178: "Socket lookup failed",
 	179: "Socket assign failed",
 	180: "Proxy redirection not supported for protocol",
+	181: "Policy denied (denylist)",
```
Nit:
```suggestion
	181: "Policy denied by denylist",
```

pchaigno reviewed (approved):
I went through all changes quickly but only reviewed the few @cilium/bpf code changes. I also made a couple tests in a GKE cluster. Good news: deny host policies work! :tada: Thanks André!

We'll have to be very clear that adding a deny policy switches all selected endpoints in default deny (and therefore denies everything) because it's very easy to shoot yourself in the foot otherwise. The consequences are a bit painful when this is a host policy...

Maybe this "deny a couple things but default allow" is even a use case to highlight in the documentation:
```yaml
  ingress:
  - fromEntities:
    - all
  ingressDeny:
  - fromCIDR:
    - 1.1.1.1/32
```
If folks don't write this the "right" way, it could quickly become painful. For example, if they use `cluster` instead of `all`, they'll end up with lots of entries in their policy maps and lots of regenerations for each new endpoint created in the cluster.

aanm committed:
test: add some e2e coverage for deny policies

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -158,7 +170,15 @@ func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
-		} else {
+		case allCpy == newKey:
+			// If we adding a deny "all" entry, then we will remove all allow
+			// entries from the map state for that direction.
```
We could also delete all deny entries in addition to the allow, as these more specific denies would be covered by the deny "all" entry.

aanm committed:
docs: add deny policies documentation

Add documentation and examples for policy denies.

Signed-off-by: André Martins <andre@cilium.io>

christarazi reviewed (approved):
:tada: Mainly looked over CI code changes this time around.

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1069,6 +1069,100 @@ that do not need it. See the `Kubernetes documentation <https://kubernetes.io/do
++--------------------------------------------------------------------------------------------+
+| **Ingress Policies Deployed to Server Pod (Yes / No)**                                     |
++---------------------+-----------------------+---------+---------+--------+--------+--------+
+|                     | Layer 7 (HTTP)        | Yes     | Yes     | Yes    | Yes    | No     |
+|                     +-----------------------+---------+---------+--------+--------+--------+
```
Does the L7 row here have any effect? Or are we just keeping it here to make sure it's known that L7 won't override Deny policy?

aanm replied:
Just to make sure the L7 won't override Deny policy.

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1069,6 +1069,100 @@ that do not need it. See the `Kubernetes documentation <https://kubernetes.io/do
+explicitly restrict certain traffic to and from a Pod.
+
+If multiple allow and deny policies are applied to the same pod, the following
+table represents the expected enforcement for that Pod:
```
IMO we should establish some baselines around the policies before we jump into a table that the reader needs to spend time trying to figure out how to interpret.

First paragraph is fine. After that I would outline that deny policies take precedence over allow policies, regardless of whether they're CNP or CCNP. Then I would outline the default-deny behaviour which is currently in a paragraph after the table (but drop the table reference). I think with those two pieces in place, the reader will be better prepared to dive into the table.

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1069,6 +1069,100 @@ that do not need it. See the `Kubernetes documentation <https://kubernetes.io/do
+
+Similarly to "allow" policies, Pods will enter default-deny mode as soon a
+single policy selects it. That is why the last column of the previous table
+denies all kind of traffic when a single deny policy is deployed.
```
Assuming we change the docs per my feedback above, most of this paragraph would go above the table. The second sentence would not make sense above the table though, but it may still be useful here to pick one of the columns and describe how to interpret the table, eg something like:

```suggestion
Similarly to "allow" policies, Pods will enter default-deny mode as soon a
single policy selects it. That is why the last column of the previous table
denies all kind of traffic when a single deny policy is deployed.
```
```suggestion
If we pick the second column in the above table, the bottom section shows the forwarding behaviour for a policy that selects curl or ping traffic between the client and server:
* Curl to port 81 is allowed because there is an allow policy on port 81, and no deny policy on that port.
* Curl to port 80 is denied because there is a deny policy on that port.
* Ping to the server is allowed because there is a Layer 3 allow policy and no deny.
```

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1069,6 +1069,100 @@ that do not need it. See the `Kubernetes documentation <https://kubernetes.io/do
+
+As opposed to allow policies, deny policies do not support: policy enforcement
+at L7, i.e., specifically denying an URL and ``toFQDNs``, i.e., specifically
+denying traffic to a specific domain name.
```
Minor nit, I don't think the first sentence fragment adds any clarity, the statement is still accurate and correct without it:

```suggestion
Deny policies do not support: policy enforcement
at L7, i.e., specifically denying an URL and ``toFQDNs``, i.e., specifically
denying traffic to a specific domain name.
```

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1069,6 +1069,100 @@ that do not need it. See the `Kubernetes documentation <https://kubernetes.io/do
+----------------------------
+
+The current known limitation is a deny policy with ``toEntities`` "world" for
+which a ``toFQDNs`` can bypass it.
```
Maybe we should file an issue for this and reference it here so if users are concerned about it, they can :+1:  and/or investigate?

joestringer replied:
`it` at the end of the sentence is unclear. Please be more specific, even if it takes a few more words to outline (1) what the behaviour is right now and (2) that this is unintended and subject to change.

joestringer commented on examples/policies/l3/entities/from_world_deny.yaml:
```diff
@@ -0,0 +1,9 @@
+metadata:
+  name: "external-lockdown"
+spec:
+  endpointSelector: {}
+  ingressDeny:
+  - fromEntities:
+    - "world"
```
Maybe I missed but is this referenced somewhere? I wonder if it should have an `ingress: -fromEntities: - all` to ensure it doesn't just drop all traffic

joestringer reviewed (approved):
I did another brief pass particularly looking at the docs, I think there's some improvements we can make before release but I'm OK to push those out to a subsequent PR.

Given the level of review we've done between Jarno, you and I across the core logic + eyes on the overall k8s API behaviour from a broader set of folks, I think we have nailed down the desired behaviour. Furthermore we've caught a bunch of issues and added testing for several areas that were previously ambiguous so I'm getting towards diminishing returns on spending more time staring at the code in this PR. Given the beta tag I think we're safe to get this in, and push it out for broader exposure, testing, and iteration on top of this base.

Are there any remaining areas you feel could do with more eyes/attention?
